test/unit/test-action-allowlist.js

```diff
@@ -0,0 +1,446 @@
+}
+
+function getActionInvocation(element, action, opt_tagOrTarget) {
+  return new ActionInvocation(
+    element,
+    action,
+    /* args */ null,
+    'source',
+    'caller',
+    'event',
+    ActionTrust.HIGH,
+    'tap',
+    opt_tagOrTarget || element.tagName
+  );
+}
+
+describes.realWin(
+  'Action allowlist on components',
+  {
+    amp: {
+      runtimeOn: true,
+      extensions: [
+        'amp-carousel',
+        'amp-form',
+        'amp-image-lightbox',
+        'amp-lightbox',
+        'amp-list',
+        'amp-selector',
+        'amp-sidebar',
+      ],
+    },
+  },
+  (env) => {
+    let action;
+    let target;
+    let spy;
+    let getDefaultActionAlias;
+
+    beforeEach(() => {
+      spy = env.sandbox.spy();
+      getDefaultActionAlias = env.sandbox.stub();
+      target = createExecElement('foo', spy, getDefaultActionAlias);
+    });
+
+    describe('with null action allowlist', () => {
+      beforeEach(() => {
+        action = new ActionService(env.ampdoc, env.win.document);
+      });
+
+      afterEach(() => {
+        action = null;
+      });
+
+      it('should allow all actions by default', () => {
+        const i = getActionInvocation(target, 'setState', 'AMP');
+        action.invoke_(i);
+        expect(spy).to.be.calledWithExactly(i);
+      });
+
+      it('should allow all actions case insensitive', () => {
+        const i = getActionInvocation(target, 'setState', 'amp');
+        action.invoke_(i);
+        expect(spy).to.be.calledWithExactly(i);
+      });
+    });
+
+    describe('with non-null action allowlist', () => {
+      beforeEach(() => {
+        action = new ActionService(env.ampdoc, env.win.document);
+        action.setWhitelist([
+          {tagOrTarget: 'AMP', method: 'pushState'},
+          {tagOrTarget: 'AMP', method: 'setState'},
+          {tagOrTarget: '*', method: 'show'},
+          {tagOrTarget: 'amp-element', method: 'defaultAction'},
+        ]);
+      });
+
+      it('should allow default actions if alias is registered default', () => {
+        // Given that 'defaultAction' is a registered default action.
+        getDefaultActionAlias.returns('defaultAction');
+        // Expect the 'activate' call to invoke it.
+        const i = getActionInvocation(target, 'activate', null);
+        action.invoke_(i);
+        expect(spy).to.be.calledWithExactly(i);
+      });
+
+      it('should allow whitelisted actions with wildcard target', () => {
+        const i = getActionInvocation(target, 'show', 'DIV');
+        action.invoke_(i);
+        expect(spy).to.be.calledWithExactly(i);
+      });
+
+      it('should not allow non-whitelisted actions', () => {
+        const i = getActionInvocation(target, 'print', 'AMP');
+        env.sandbox.stub(action, 'error_');
+        expect(action.invoke_(i)).to.be.null;
+        expect(action.error_).to.be.calledWith(
+          '"AMP.print" is not whitelisted ' +
+            '[{"tagOrTarget":"AMP","method":"pushState"},' +
```

Comment:
Yikes, is that what the error message looks like today? Optional suggestion to make it less ugly.

Reply:
Ugly but helpful. Do you suggest making it less ugly by removing some info or parsing it into a more readable state? It may be better to have the `isActionWhitelisted` method throw errors inline with missed conditions instead of having the invocation method throw a blanket error on an unfavorable result. Still I'm going to leave it out of the scope of this change if that's OK.

Reply:
Yea I was just thinking of printing "AMP.pushState" instead of "{"tagOrTarget":"AMP","method":"pushState"}", for example. Fine to leave out.